WIP: job-exec/sdexec: enable resource control properties MemoryMax, MemoryHigh, AllowedCPUs, etc.

chu11 opened the pull request:
This PR adds initial support to configure systemd properties via `libsdprocess`.

Long term we will want to support many potential configurations, such as cgroups configurations or hard/soft limits.

As an initial implementation for testing, I added support for `CPUAffinity` settings and had the `sdexec` module set `CPUAffinity` for a process based on the resources given to it in R.

Note, this is built on top of #4070.

Review considerations:

- in order to convert from "cores" to "processing units", did some `hwloc` hackery, hope I did it right :-)

- the "cores" -> "processing units" calculation is done in `sdexec` instead of `libsdprocess`.  Maybe that's a bad idea.  It was done mostly b/c systemd expects the caller to do that.  So I similarly left it like that for callers of `libsdprocess`.

Other note:

I initially tried to implement `AllowedCPUs` (cpu containment), but that is only supported in RHEL8 with cgroups v2.  cgroups v1 is the default on RHEL8 and cgroups v2 is a tech preview.  Something to think about for the future and what we'd want to do.

chu11 commented:
re-pushed, rebasing on PR #4070 and fixing up some #4278 conflicts

chu11 commented:
just rebasing on #4070

chu11 commented:
rebased on master, since it'd been awhile

grondo commented on src/modules/job-exec/sdexec.c:
```diff
@@ -191,6 +214,218 @@ static void set_stdlog (struct sdexec *se,
+                goto error;
+            }
+            hwloc_bitmap_set (puset, core->children[j]->os_index);
+        }
```
This function is largely a copy of what is done in `src/shell/affinity.c` `shell_affinity_get_cpuset()`.
Perhaps not necessary to copy that function, nor to create a libutil function at this point, but a simplification from the existing implementation would be to use `hwloc_bitmap_or ()` on `core->cpuset` instead of by hand setting the bit of each core child. (This lets you drop this entire loop).

grondo reviewed (commented):
It is unfortunate how much code is required here just to support the example `CPUAffinity` property, which sadly is not all that useful for real jobs.

I wonder if we need a plan to boot fluke with `systemd.unified_cgroup_hierarchy=1` and see if we can get `MaxMemory` and `AllowedCPUs` working before we move forward with this feature.

Also, in the end, it would be nice if sysadmins could configure any supported resource control property for jobs when the sdexec implementation is being used. However, looking at the amount of "setup" required just for `CPUAffinity`, maybe it isn't even possible to allow "general" properties to be configured? What is your feeling on that? So if sysadmins wanted to configure a new property for all jobs, it would require a bunch of code and a new version of flux-core before that could happen?

I'm just thinking out loud about the maintenance burden this feature might incur. Maybe it would be easier in the long run to focus on allowing the IMP to setup all the cgroups values before a job is started.

Mainly asking for your feedback on these thoughts for now @chu11!

chu11 commented:
> Also, in the end, it would be nice if sysadmins could configure any supported resource control property for jobs when the sdexec implementation is being used. However, looking at the amount of "setup" required just for CPUAffinity, maybe it isn't even possible to allow "general" properties to be configured? What is your feeling on that? So if sysadmins wanted to configure a new property for all jobs, it would require a bunch of code and a new version of flux-core before that could happen?

I suppose it might technically be possible, although it would probably be a difficult configuration involving bitmasks.  i.e. read from the config and transfer directly into sd_bus API calls.

This was admittedly surprised me when working on this.  I had thought things like "CPUAffinity=1,4" or whatever would be translated inside the sd-bus library, but its not.  In systemd's code, there's a mountain of code in `systemd-run` just for these conversions before calling the sd_bus calls.

As we add things of course there can be some code re-use (CPUAffinity and AllowedCPUs are virtually the same).  And any properties that are on/off are much easier to do of course.

> I'm just thinking out loud about the maintenance burden this feature might incur. Maybe it would be easier in the long run to focus on allowing the IMP to setup all the cgroups values before a job is started.

Do you mean "hand configuration" by setting things up in `/sys`?

Even though it would require releases to support new systemd configs, I think this would be a win "long term".  We probably don't want to deal with the minutiae on configuring cgroups by hand?  and how things could change with different cgroups versions?

chu11 commented on src/common/libsdprocess/sdprocess.c:
```diff
@@ -277,12 +279,99 @@ static sdprocess_t *sdprocess_create (flux_t *h,
+    return -1;
+}
+
+static int set_property_CPUAffinity (sdprocess_t *sdp,
+                                     sd_bus_message *m,
+                                     const char *property,
+                                     const char *value)
```
yup, I think I was thinking a little bit farther ahead when this could be used for multiple properties.  But you're right, unneeded right now.

grondo commented:
> As we add things of course there can be some code re-use (CPUAffinity and AllowedCPUs are virtually the same). And any properties that are on/off are much easier to do of course.

Good point, though CPUAffinity isn't a real use case so there perhaps is no point supporting it once/if AllowedCPUs is working. Do you know if MaxMemory works with the cgroups v1 hierarchy?

> Do you mean "hand configuration" by setting things up in /sys?

Yeah, or using something like libcgroup or similar library if a useful one exists. 

> Even though it would require releases to support new systemd configs, I think this would be a win "long term". We probably don't want to deal with the minutiae on configuring cgroups by hand? 

We may be able to use an existing library. Also, if we can provide the functionality to configure cgroups directly in Flux, then we can use plugins for the per-job configuration, allowing sites and users to develop their own plugins and not push all this specific code directly into flux-core, nor require a release every time some new controller or a small change in behavior is required. Additionally, if we have at least some support for cgroups in Flux, we could potentially have cgroups support even for single-user instances, since we can delegate the system instance job cgroup hierarchy to the new instance owner.

And besides, it looks like we're going to have to do the work ourselves since all the useful systemd properties (e.g. AllowedCPUs and AllowedMemoryNodes) don't seem to really be supported in RHEL8 without a configuration change (and we don't know yet if we'll be able to request that change)

I'm not arguing against the systemd approach, I'm just wondering if it makes more sense to direct our effort elsewhere for now. I guess the first thing to do is to ask if we can flip fluke to the unified cgroups hierarchy.

chu11 commented:
re-pushed with fixes per comments above, went ahead and squashed since they were pretty minor changes

chu11 commented:
just added some commits to add `AllowedCPUs` support to `libsdprocess`.  I left `CPUAffinity` support in there b/c it was sort of nice to leave the unit tests for `libsdprocess` in there.  They are easily programmable and testable w/ `sched_getaffinity()` and such.  vs. `AllowedCPUs` where I'd have to do a bunch of digging through `/sys/fs/cgroups`, read & parse files, etc. and that's a pain.  So the only tests added for `AllowedCPUs` is through sharness tests and the `sdexec` module.

However, since we don't need `CPUAffinity` short term, it would be understandable if we wanted to axe it as well.  We can debate that.

note, that I intend to eventually add MemoryMax and similar support in a separate PR, since those properties will be quite different these CPU based ones.

chu11 commented:
re-pushed with MemoryHigh and MemoryMax from PR #4367 in here.  Re-worked some commits b/c few things could be merged together (e.g. two documentation commits now are one).

grondo commented:
Didn't forget about this one @chu11 - on my list to look at this week.

chu11 commented:
rebased & re-pushed, was trying to figure out why the `libsdprocess` `test_parse.t` was segfaulting on a few builders.

Eventually found it, I didn't escape the percent sign in some output messages.  And in the depths of tap's `ok()` macros, it eventually borked.  Wasn't immediately obvious b/c you're not using a printf-ish function.  Replace `%` with "percent" and added a few extra unit tests to boot.

chu11 commented:
rebased & re-pushed, just fixing up a conflict

chu11 commented:
per conversation in #5348 we will keep this here for "documentation" before a number of other features get pulled in at a later time.  So I'll this as WIP indefinitely.

garlick commented:
Closing this as we have a different framework now for setting resource control properties in systemd-launched jobs.  (It'll still be here if we need to revive the work on AllowedCPUs for example.
